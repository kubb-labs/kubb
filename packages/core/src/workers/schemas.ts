import { z } from 'zod'

import type { ZodStringDef, ZodType } from 'zod'
import type { KubbFile } from '../FileManager.ts'

export const importSchema = z.object({
  name: z.union([z.string(), z.array(z.string())]),
  path: z.string(),
  isTypeOnly: z.boolean().optional(),
  isNameSpace: z.boolean().optional(),
  root: z.string().optional(),
})

export const exportSchema = z.object({
  name: z.union([z.string(), z.array(z.string())]).optional(),
  path: z.string(),
  isTypeOnly: z.boolean().optional(),
  asAlias: z.boolean().optional(),
})

export const fileSchema = z.object({
  id: z.string().optional(),
  baseName: z.string() as ZodType<KubbFile.BaseName, ZodStringDef>,
  path: z.string(),
  source: z.string(),
  imports: z.array(importSchema).optional(),
  exports: z.array(exportSchema).optional(),
  override: z.boolean().optional(),
  meta: z.object({}).optional(),
  env: z.object({}).optional(),
})
