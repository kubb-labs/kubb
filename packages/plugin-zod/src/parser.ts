import transformers from '@kubb/core/transformers'
import type { SchemaObject } from '@kubb/oas'

import type { Schema, SchemaKeywordBase, SchemaMapper } from '@kubb/plugin-oas'
import { isKeyword, SchemaGenerator, type SchemaKeywordMapper, type SchemaTree, schemaKeywords } from '@kubb/plugin-oas'

//TODO add zodKeywordMapper as function that returns 3 versions: v3, v4 and v4 mini, this can also be used to have the custom mapping(see object type)
// also include shouldCoerce

const zodKeywordMapper = {
  any: () => 'z.any()',
  unknown: () => 'z.unknown()',
  void: () => 'z.void()',
  number: (coercion?: boolean, min?: number, max?: number, exclusiveMinimum?: number, exclusiveMaximum?: number) => {
    return [
      coercion ? 'z.coerce.number()' : 'z.number()',
      min !== undefined ? `.min(${min})` : undefined,
      max !== undefined ? `.max(${max})` : undefined,
      exclusiveMinimum !== undefined ? `.gt(${exclusiveMinimum})` : undefined,
      exclusiveMaximum !== undefined ? `.lt(${exclusiveMaximum})` : undefined,
    ]
      .filter(Boolean)
      .join('')
  },
  integer: (coercion?: boolean, min?: number, max?: number, version: '3' | '4' = '3', exclusiveMinimum?: number, exclusiveMaximum?: number) => {
    return [
      coercion ? 'z.coerce.number().int()' : version === '4' ? 'z.int()' : 'z.number().int()',
      min !== undefined ? `.min(${min})` : undefined,
      max !== undefined ? `.max(${max})` : undefined,
      exclusiveMinimum !== undefined ? `.gt(${exclusiveMinimum})` : undefined,
      exclusiveMaximum !== undefined ? `.lt(${exclusiveMaximum})` : undefined,
    ]
      .filter(Boolean)
      .join('')
  },
  interface: (value?: string, strict?: boolean) => {
    if (strict) {
      return `z.strictInterface({
    ${value}
    })`
    }
    return `z.interface({
    ${value}
    })`
  },
  object: (value?: string, strict?: boolean, version: '3' | '4' = '3') => {
    if (version === '4' && strict) {
      return `z.strictObject({
    ${value}
    })`
    }

    if (strict) {
      return `z.object({
    ${value}
    }).strict()`
    }

    return `z.object({
    ${value}
    })`
  },
  string: (coercion?: boolean, min?: number, max?: number) => {
    return [coercion ? 'z.coerce.string()' : 'z.string()', min !== undefined ? `.min(${min})` : undefined, max !== undefined ? `.max(${max})` : undefined]
      .filter(Boolean)
      .join('')
  },
  //support for discriminatedUnion
  boolean: () => 'z.boolean()',
  undefined: () => 'z.undefined()',
  nullable: (value?: string) => {
    if (value) {
      return `z.nullable(${value})`
    }
    return '.nullable()'
  },
  null: () => 'z.null()',
  nullish: (value?: string) => {
    if (value) {
      return `z.nullish(${value})`
    }
    return '.nullish()'
  },
  array: (items: string[] = [], min?: number, max?: number, unique?: boolean) => {
    return [
      `z.array(${items?.join('')})`,
      min !== undefined ? `.min(${min})` : undefined,
      max !== undefined ? `.max(${max})` : undefined,
      unique ? `.refine(items => new Set(items).size === items.length, { message: "Array entries must be unique" })` : undefined,
    ]
      .filter(Boolean)
      .join('')
  },
  tuple: (items: string[] = []) => `z.tuple([${items?.join(', ')}])`,
  enum: (items: string[] = []) => `z.enum([${items?.join(', ')}])`,
  union: (items: string[] = []) => `z.union([${items?.join(', ')}])`,
  const: (value?: string | number | boolean) => `z.literal(${value ?? ''})`,
  /**
   * ISO 8601
   */
  datetime: (offset = false, local = false, version: '3' | '4' = '3') => {
    if (offset) {
      return version === '4' ? `z.iso.datetime({ offset: ${offset} })` : `z.string().datetime({ offset: ${offset} })`
    }

    if (local) {
      return version === '4' ? `z.iso.datetime({ local: ${local} })` : `z.string().datetime({ local: ${local} })`
    }

    return 'z.string().datetime()'
  },
  /**
   * Type `'date'` Date
   * Type `'string'` ISO date format (YYYY-MM-DD)
   * @default ISO date format (YYYY-MM-DD)
   */
  date: (type: 'date' | 'string' = 'string', coercion?: boolean, version: '3' | '4' = '3') => {
    if (type === 'string') {
      return version === '4' ? 'z.iso.date()' : 'z.string().date()'
    }

    if (coercion) {
      return 'z.coerce.date()'
    }

    return 'z.date()'
  },
  /**
   * Type `'date'` Date
   * Type `'string'` ISO time format (HH:mm:ss[.SSSSSS])
   * @default ISO time format (HH:mm:ss[.SSSSSS])
   */
  time: (type: 'date' | 'string' = 'string', coercion?: boolean, version: '3' | '4' = '3') => {
    if (type === 'string') {
      return version === '4' ? 'z.iso.time()' : 'z.string().time()'
    }

    if (coercion) {
      return 'z.coerce.date()'
    }

    return 'z.date()'
  },
  uuid: (coercion?: boolean, version: '3' | '4' = '3', min?: number, max?: number) => {
    return [
      coercion ? 'z.coerce.string().uuid()' : version === '4' ? 'z.uuid()' : 'z.string().uuid()',
      min !== undefined ? `.min(${min})` : undefined,
      max !== undefined ? `.max(${max})` : undefined,
    ]
      .filter(Boolean)
      .join('')
  },
  url: (coercion?: boolean, version: '3' | '4' = '3', min?: number, max?: number) => {
    return [
      coercion ? 'z.coerce.string().url()' : version === '4' ? 'z.url()' : 'z.string().url()',
      min !== undefined ? `.min(${min})` : undefined,
      max !== undefined ? `.max(${max})` : undefined,
    ]
      .filter(Boolean)
      .join('')
  },
  default: (value?: string | number | true | object) => {
    if (typeof value === 'object') {
      return '.default({})'
    }
    return `.default(${value ?? ''})`
  },
  and: (items: string[] = []) => items?.map((item) => `.and(${item})`).join(''),
  describe: (value = '') => `.describe(${value})`,
  max: undefined,
  min: undefined,
  optional: (value?: string) => {
    if (value) {
      return `z.optional(${value})`
    }
    return '.optional()'
  },
  matches: (value = '', coercion?: boolean) => (coercion ? `z.coerce.string().regex(${value})` : `z.string().regex(${value})`),
  email: (coercion?: boolean, version: '3' | '4' = '3', min?: number, max?: number) => {
    return [
      coercion ? 'z.coerce.string().email()' : version === '4' ? 'z.email()' : 'z.string().email()',
      min !== undefined ? `.min(${min})` : undefined,
      max !== undefined ? `.max(${max})` : undefined,
    ]
      .filter(Boolean)
      .join('')
  },
  firstName: undefined,
  lastName: undefined,
  password: undefined,
  phone: undefined,
  readOnly: undefined,
  writeOnly: undefined,
  ref: (value?: string, version: '3' | '4' = '3') => {
    if (!value) {
      return undefined
    }

    return version === '4' ? value : `z.lazy(() => ${value})`
  },
  blob: () => 'z.instanceof(File)',
  deprecated: undefined,
  example: undefined,
  schema: undefined,
  catchall: (value?: string) => (value ? `.catchall(${value})` : undefined),
  name: undefined,
  exclusiveMinimum: undefined,
  exclusiveMaximum: undefined,
} satisfies SchemaMapper<string | null | undefined>

/**
 * @link based on https://github.com/cellular/oazapfts/blob/7ba226ebb15374e8483cc53e7532f1663179a22c/src/codegen/generate.ts#L398
 */

export function sort(items?: Schema[]): Schema[] {
  const order: string[] = [
    schemaKeywords.string,
    schemaKeywords.datetime,
    schemaKeywords.date,
    schemaKeywords.time,
    schemaKeywords.tuple,
    schemaKeywords.number,
    schemaKeywords.object,
    schemaKeywords.enum,
    schemaKeywords.url,
    schemaKeywords.email,
    schemaKeywords.firstName,
    schemaKeywords.lastName,
    schemaKeywords.password,
    schemaKeywords.matches,
    schemaKeywords.uuid,
    schemaKeywords.null,
    schemaKeywords.min,
    schemaKeywords.max,
    schemaKeywords.default,
    schemaKeywords.describe,
    schemaKeywords.optional,
    schemaKeywords.nullable,
    schemaKeywords.nullish,
  ]

  if (!items) {
    return []
  }

  return transformers.orderBy(items, [(v) => order.indexOf(v.keyword)], ['asc'])
}

const shouldCoerce = (coercion: ParserOptions['coercion'] | undefined, type: 'dates' | 'strings' | 'numbers'): boolean => {
  if (coercion === undefined) {
    return false
  }
  if (typeof coercion === 'boolean') {
    return coercion
  }

  return !!coercion[type]
}

type ParserOptions = {
  name: string
  typeName?: string
  description?: string
  keysToOmit?: string[]
  mapper?: Record<string, string>
  coercion?: boolean | { dates?: boolean; strings?: boolean; numbers?: boolean }
  wrapOutput?: (opts: { output: string; schema: any }) => string | undefined
  rawSchema: SchemaObject
  version: '3' | '4'
}

export function parse({ parent, current, name, siblings }: SchemaTree, options: ParserOptions): string | undefined {
  const value = zodKeywordMapper[current.keyword as keyof typeof zodKeywordMapper]

  // Early exit: if siblings contain both matches and ref â†’ skip matches entirely
  const hasMatches = siblings.some((it) => isKeyword(it, schemaKeywords.matches))
  const hasRef = siblings.some((it) => isKeyword(it, schemaKeywords.ref))

  if (hasMatches && hasRef && isKeyword(current, schemaKeywords.matches)) {
    return undefined // strip matches
  }

  if (!value) {
    return undefined
  }

  if (isKeyword(current, schemaKeywords.union)) {
    // zod union type needs at least 2 items
    if (Array.isArray(current.args) && current.args.length === 1) {
      return parse({ parent, name: name, current: current.args[0] as Schema, siblings }, options)
    }
    if (Array.isArray(current.args) && !current.args.length) {
      return ''
    }

    return zodKeywordMapper.union(
      sort(current.args)
        .map((schema, _index, siblings) => parse({ parent: current, name: name, current: schema, siblings }, options))
        .filter(Boolean),
    )
  }

  if (isKeyword(current, schemaKeywords.and)) {
    const items = sort(current.args)
      .filter((schema: Schema) => {
        return ![schemaKeywords.optional, schemaKeywords.describe].includes(schema.keyword as typeof schemaKeywords.describe)
      })
      .map((schema: Schema, _index, siblings) => parse({ parent: current, name: name, current: schema, siblings }, options))
      .filter(Boolean)

    return `${items.slice(0, 1)}${zodKeywordMapper.and(items.slice(1))}`
  }

  if (isKeyword(current, schemaKeywords.array)) {
    return zodKeywordMapper.array(
      sort(current.args.items)
        .map((schemas, _index, siblings) => parse({ parent: current, name: name, current: schemas, siblings }, options))
        .filter(Boolean),
      current.args.min,
      current.args.max,
      current.args.unique,
    )
  }

  if (isKeyword(current, schemaKeywords.enum)) {
    if (current.args.asConst) {
      if (current.args.items.length === 1) {
        const child = {
          keyword: schemaKeywords.const,
          args: current.args.items[0],
        }
        return parse({ parent: current, name: name, current: child, siblings: [child] }, options)
      }

      return zodKeywordMapper.union(
        current.args.items
          .map((schema) => ({
            keyword: schemaKeywords.const,
            args: schema,
          }))
          .map((schema, _index, siblings) => {
            return parse({ parent: current, name: name, current: schema, siblings }, options)
          })
          .filter(Boolean),
      )
    }

    return zodKeywordMapper.enum(
      current.args.items.map((schema) => {
        if (schema.format === 'boolean') {
          return transformers.stringify(schema.value)
        }

        if (schema.format === 'number') {
          return transformers.stringify(schema.value)
        }
        return transformers.stringify(schema.value)
      }),
    )
  }

  if (isKeyword(current, schemaKeywords.ref)) {
    return zodKeywordMapper.ref(current.args?.name, options.version)
  }

  if (isKeyword(current, schemaKeywords.object)) {
    const propertyEntries = Object.entries(current.args?.properties || {}).filter((item) => {
      const schema = item[1]
      return schema && typeof schema.map === 'function'
    })

    const properties = propertyEntries
      .map(([name, schemas]) => {
        const nameSchema = schemas.find((it) => it.keyword === schemaKeywords.name) as SchemaKeywordMapper['name']
        const isNullable = schemas.some((it) => isKeyword(it, schemaKeywords.nullable))
        const isNullish = schemas.some((it) => isKeyword(it, schemaKeywords.nullish))
        const isOptional = schemas.some((it) => isKeyword(it, schemaKeywords.optional))

        const mappedName = nameSchema?.args || name

        // custom mapper(pluginOptions)
        if (options.mapper?.[mappedName]) {
          return `"${name}": ${options.mapper?.[mappedName]}`
        }

        const baseSchemaOutput = sort(schemas)
          .filter((schema) => {
            return !isKeyword(schema, schemaKeywords.optional) && !isKeyword(schema, schemaKeywords.nullable) && !isKeyword(schema, schemaKeywords.nullish)
          })
          .map((schema) => parse({ parent: current, name, current: schema, siblings: schemas }, options))
          .filter(Boolean)
          .join('')

        const objectValue = options.wrapOutput
          ? options.wrapOutput({ output: baseSchemaOutput, schema: options.rawSchema?.properties?.[name] }) || baseSchemaOutput
          : baseSchemaOutput

        if (options.version === '4' && SchemaGenerator.find(schemas, schemaKeywords.ref)) {
          // both optional and nullable
          if (isNullish) {
            return `get "${name}"(){
                return ${zodKeywordMapper.nullish(objectValue)}
              }`
          }

          // undefined
          if (isOptional) {
            return `get "${name}"(){
                return ${zodKeywordMapper.optional(objectValue)}
              }`
          }

          // null
          if (isNullable) {
            return `get "${name}"(){
                return ${zodKeywordMapper.nullable(objectValue)}
              }`
          }

          return `get "${name}"(){
                return ${objectValue}
              }`
        }

        // both optional and nullable
        if (isNullish) {
          return `"${name}": ${objectValue}${zodKeywordMapper.nullish()}`
        }

        // undefined
        if (isOptional) {
          return `"${name}": ${zodKeywordMapper.optional(objectValue)}`
        }

        // null
        if (isNullable) {
          return `"${name}": ${zodKeywordMapper.nullable(objectValue)}`
        }

        return `"${name}": ${objectValue}`
      })
      .join(',\n')

    const additionalProperties = current.args?.additionalProperties?.length
      ? current.args.additionalProperties
          .map((schema, _index, siblings) => parse({ parent: current, name: name, current: schema, siblings }, options))
          .filter(Boolean)
          .join('')
      : undefined

    const text = [
      zodKeywordMapper.object(properties, current.args?.strict, options.version),
      additionalProperties ? zodKeywordMapper.catchall(additionalProperties) : undefined,
    ].filter(Boolean)

    return text.join('')
  }

  if (isKeyword(current, schemaKeywords.tuple)) {
    return zodKeywordMapper.tuple(
      current.args.items.map((schema, _index, siblings) => parse({ parent: current, name: name, current: schema, siblings }, options)).filter(Boolean),
    )
  }

  if (isKeyword(current, schemaKeywords.const)) {
    if (current.args.format === 'number' && current.args.value !== undefined) {
      return zodKeywordMapper.const(Number(current.args.value))
    }

    if (current.args.format === 'boolean' && current.args.value !== undefined) {
      return zodKeywordMapper.const(current.args.value)
    }
    return zodKeywordMapper.const(transformers.stringify(current.args.value))
  }

  if (isKeyword(current, schemaKeywords.matches)) {
    if (current.args) {
      return zodKeywordMapper.matches(transformers.toRegExpString(current.args, null), shouldCoerce(options.coercion, 'strings'))
    }
  }

  if (isKeyword(current, schemaKeywords.default)) {
    if (current.args) {
      return zodKeywordMapper.default(current.args)
    }
  }

  if (isKeyword(current, schemaKeywords.describe)) {
    if (current.args) {
      return zodKeywordMapper.describe(transformers.stringify(current.args.toString()))
    }
  }

  if (isKeyword(current, schemaKeywords.string)) {
    const minSchema = SchemaGenerator.find(siblings, schemaKeywords.min)
    const maxSchema = SchemaGenerator.find(siblings, schemaKeywords.max)

    return zodKeywordMapper.string(shouldCoerce(options.coercion, 'strings'), minSchema?.args, maxSchema?.args)
  }

  if (isKeyword(current, schemaKeywords.uuid)) {
    return zodKeywordMapper.uuid(shouldCoerce(options.coercion, 'strings'), options.version)
  }

  if (isKeyword(current, schemaKeywords.email)) {
    const minSchema = SchemaGenerator.find(siblings, schemaKeywords.min)
    const maxSchema = SchemaGenerator.find(siblings, schemaKeywords.max)

    return zodKeywordMapper.email(shouldCoerce(options.coercion, 'strings'), options.version, minSchema?.args, maxSchema?.args)
  }

  if (isKeyword(current, schemaKeywords.url)) {
    const minSchema = SchemaGenerator.find(siblings, schemaKeywords.min)
    const maxSchema = SchemaGenerator.find(siblings, schemaKeywords.max)

    return zodKeywordMapper.url(shouldCoerce(options.coercion, 'strings'), options.version, minSchema?.args, maxSchema?.args)
  }

  if (isKeyword(current, schemaKeywords.number)) {
    const minSchema = SchemaGenerator.find(siblings, schemaKeywords.min)
    const maxSchema = SchemaGenerator.find(siblings, schemaKeywords.max)

    const exclusiveMinimumSchema = SchemaGenerator.find(siblings, schemaKeywords.exclusiveMinimum)
    const exclusiveMaximumSchema = SchemaGenerator.find(siblings, schemaKeywords.exclusiveMaximum)
    return zodKeywordMapper.number(
      shouldCoerce(options.coercion, 'numbers'),
      minSchema?.args,
      maxSchema?.args,
      exclusiveMinimumSchema?.args,
      exclusiveMaximumSchema?.args,
    )
  }

  if (isKeyword(current, schemaKeywords.integer)) {
    const minSchema = SchemaGenerator.find(siblings, schemaKeywords.min)
    const maxSchema = SchemaGenerator.find(siblings, schemaKeywords.max)

    const exclusiveMinimumSchema = SchemaGenerator.find(siblings, schemaKeywords.exclusiveMinimum)
    const exclusiveMaximumSchema = SchemaGenerator.find(siblings, schemaKeywords.exclusiveMaximum)
    return zodKeywordMapper.integer(
      shouldCoerce(options.coercion, 'numbers'),
      minSchema?.args,
      maxSchema?.args,
      options.version,
      exclusiveMinimumSchema?.args,
      exclusiveMaximumSchema?.args,
    )
  }

  if (isKeyword(current, schemaKeywords.datetime)) {
    return zodKeywordMapper.datetime(current.args.offset, current.args.local, options.version)
  }

  if (isKeyword(current, schemaKeywords.date)) {
    return zodKeywordMapper.date(current.args.type, shouldCoerce(options.coercion, 'dates'), options.version)
  }

  if (isKeyword(current, schemaKeywords.time)) {
    return zodKeywordMapper.time(current.args.type, shouldCoerce(options.coercion, 'dates'), options.version)
  }

  if (current.keyword in zodKeywordMapper && 'args' in current) {
    const value = zodKeywordMapper[current.keyword as keyof typeof zodKeywordMapper] as (typeof zodKeywordMapper)['const']

    return value((current as SchemaKeywordBase<unknown>).args as any)
  }

  if (current.keyword in zodKeywordMapper) {
    return value()
  }

  return undefined
}
