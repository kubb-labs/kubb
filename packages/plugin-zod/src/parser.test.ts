import { schemas } from '../../plugin-oas/mocks/schemas.ts'

import * as parserZod from './parser.ts'

describe('zod parse', () => {
  test.each(schemas.basic)('$name', ({ name, schema }) => {
    const text = parserZod.parse({ parent: undefined, current: schema, siblings: [schema] }, { name, rawSchema: {} })
    expect(text).toMatchSnapshot()
  })
})
