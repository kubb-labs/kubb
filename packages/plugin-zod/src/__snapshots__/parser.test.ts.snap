// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`zod parse > 'and' 1`] = `"z.string().and(z.number())"`;

exports[`zod parse > 'and' v4 1`] = `"z.string().and(z.number())"`;

exports[`zod parse > 'any' 1`] = `"z.any()"`;

exports[`zod parse > 'any' v4 1`] = `"z.any()"`;

exports[`zod parse > 'array' 1`] = `"z.array(z.union([z.string(), z.number()]))"`;

exports[`zod parse > 'array' v4 1`] = `"z.array(z.union([z.string(), z.number()]))"`;

exports[`zod parse > 'arrayAdvanced' 1`] = `"z.array(z.number().min(1).max(10)).min(3).max(10)"`;

exports[`zod parse > 'arrayAdvanced' v4 1`] = `"z.array(z.number().min(1).max(10)).min(3).max(10)"`;

exports[`zod parse > 'arrayEmpty' 1`] = `"z.array()"`;

exports[`zod parse > 'arrayEmpty' v4 1`] = `"z.array()"`;

exports[`zod parse > 'arrayRef' 1`] = `"z.array(z.lazy(() => Pet))"`;

exports[`zod parse > 'arrayRef' v4 1`] = `"z.array(Pet)"`;

exports[`zod parse > 'arrayRegex' 1`] = `"z.array(z.string().regex(/^[a-zA-Z0-9]{1,13}$/)).min(3).max(10)"`;

exports[`zod parse > 'arrayRegex' v4 1`] = `"z.array(z.string().regex(/^[a-zA-Z0-9]{1,13}$/)).min(3).max(10)"`;

exports[`zod parse > 'blob' 1`] = `"z.instanceof(File)"`;

exports[`zod parse > 'blob' v4 1`] = `"z.instanceof(File)"`;

exports[`zod parse > 'boolean' 1`] = `"z.boolean()"`;

exports[`zod parse > 'boolean' v4 1`] = `"z.boolean()"`;

exports[`zod parse > 'catchall' 1`] = `
"z.object({
    
    }).catchall(z.lazy(() => Pet))"
`;

exports[`zod parse > 'catchall' v4 1`] = `
"z.object({
    
    }).catchall(Pet)"
`;

exports[`zod parse > 'const' 1`] = `"z.literal("")"`;

exports[`zod parse > 'const' v4 1`] = `"z.literal("")"`;

exports[`zod parse > 'date' 1`] = `"z.string().date()"`;

exports[`zod parse > 'date' v4 1`] = `"z.iso.date()"`;

exports[`zod parse > 'datetime' 1`] = `"z.string().datetime()"`;

exports[`zod parse > 'datetime' v4 1`] = `"z.string().datetime()"`;

exports[`zod parse > 'default' 1`] = `".default()"`;

exports[`zod parse > 'default' 2`] = `".default(default)"`;

exports[`zod parse > 'default' v4 1`] = `".default()"`;

exports[`zod parse > 'default' v4 2`] = `".default(default)"`;

exports[`zod parse > 'enum' 1`] = `"z.enum(["A", "B", "C", "2"])"`;

exports[`zod parse > 'enum' v4 1`] = `"z.enum(["A", "B", "C", "2"])"`;

exports[`zod parse > 'enumLiteralBoolean' 1`] = `"z.union([z.literal(true), z.literal(false)])"`;

exports[`zod parse > 'enumLiteralBoolean' v4 1`] = `"z.union([z.literal(true), z.literal(false)])"`;

exports[`zod parse > 'integer' 1`] = `"z.number().int()"`;

exports[`zod parse > 'integer' v4 1`] = `"z.int()"`;

exports[`zod parse > 'matches' 1`] = `"z.string().regex(/^[A-Z]{2}$/)"`;

exports[`zod parse > 'matches' v4 1`] = `"z.string().regex(/^[A-Z]{2}$/)"`;

exports[`zod parse > 'matchesReg' 1`] = `"z.string().regex(/node_modules/)"`;

exports[`zod parse > 'matchesReg' v4 1`] = `"z.string().regex(/node_modules/)"`;

exports[`zod parse > 'max' 1`] = `undefined`;

exports[`zod parse > 'max' v4 1`] = `undefined`;

exports[`zod parse > 'min' 1`] = `undefined`;

exports[`zod parse > 'min' v4 1`] = `undefined`;

exports[`zod parse > 'nullable' 1`] = `".nullable()"`;

exports[`zod parse > 'nullable' v4 1`] = `".nullable()"`;

exports[`zod parse > 'nullableAdditionalProperties' 1`] = `
"z.object({
    
    }).catchall(z.string().nullable())"
`;

exports[`zod parse > 'nullableAdditionalProperties' v4 1`] = `
"z.object({
    
    }).catchall(z.string().nullable())"
`;

exports[`zod parse > 'number' 1`] = `"z.number()"`;

exports[`zod parse > 'number' v4 1`] = `"z.number()"`;

exports[`zod parse > 'object' 1`] = `
"z.object({
    "firstName": z.string().min(2),
"address": z.nullable(z.string().describe("Your address"))
    })"
`;

exports[`zod parse > 'object' v4 1`] = `
"z.object({
    "firstName": z.string().min(2),
"address": z.nullable(z.string().describe("Your address"))
    })"
`;

exports[`zod parse > 'objectAnd' 1`] = `
"z.object({
    "firstName": z.string().min(2).default(test),
"age": z.number().min(3).default(2),
"address": z.nullable(z.string().and(z.number()).describe("Your address"))
    })"
`;

exports[`zod parse > 'objectAnd' v4 1`] = `
"z.object({
    "firstName": z.string().min(2).default(test),
"age": z.number().min(3).default(2),
"address": z.nullable(z.string().and(z.number()).describe("Your address"))
    })"
`;

exports[`zod parse > 'objectArray' 1`] = `
"z.object({
    "ids": z.array(z.string().regex(/^[a-zA-Z0-9]{1,13}$/)).min(3).max(10)
    })"
`;

exports[`zod parse > 'objectArray' v4 1`] = `
"z.object({
    "ids": z.array(z.string().regex(/^[a-zA-Z0-9]{1,13}$/)).min(3).max(10)
    })"
`;

exports[`zod parse > 'objectArrayObject' 1`] = `
"z.object({
    "ids": z.array(z.object({
    "enum": z.enum(["A", "B", "C", "2"])
    })).min(3).max(10)
    })"
`;

exports[`zod parse > 'objectArrayObject' v4 1`] = `
"z.object({
    "ids": z.array(z.object({
    "enum": z.enum(["A", "B", "C", "2"])
    })).min(3).max(10)
    })"
`;

exports[`zod parse > 'objectDates' 1`] = `
"z.object({
    "dateTime": z.string().datetime({ offset: true }),
"date": z.string().date(),
"time": z.string().time(),
"nativeDate": z.date()
    })"
`;

exports[`zod parse > 'objectDates' v4 1`] = `
"z.object({
    "dateTime": z.iso.datetime({ offset: true }),
"date": z.iso.date(),
"time": z.iso.time(),
"nativeDate": z.date()
    })"
`;

exports[`zod parse > 'objectEmpty' 1`] = `
"z.object({
    
    })"
`;

exports[`zod parse > 'objectEmpty' v4 1`] = `
"z.object({
    
    })"
`;

exports[`zod parse > 'objectEnum' 1`] = `
"z.object({
    "version": z.enum(["A", "B", "C", "2"]).describe("Your address")
    })"
`;

exports[`zod parse > 'objectEnum' v4 1`] = `
"z.object({
    "version": z.enum(["A", "B", "C", "2"]).describe("Your address")
    })"
`;

exports[`zod parse > 'objectObjectEnum' 1`] = `
"z.object({
    "prop1": z.object({
    "prop2": z.enum(["A", "B"])
    })
    })"
`;

exports[`zod parse > 'objectObjectEnum' v4 1`] = `
"z.object({
    "prop1": z.object({
    "prop2": z.enum(["A", "B"])
    })
    })"
`;

exports[`zod parse > 'objectOptional' 1`] = `
"z.object({
    "firstName": z.optional(z.string().min(2)),
"address": z.nullable(z.string().describe("Your address"))
    })"
`;

exports[`zod parse > 'objectOptional' v4 1`] = `
"z.object({
    "firstName": z.optional(z.string().min(2)),
"address": z.nullable(z.string().describe("Your address"))
    })"
`;

exports[`zod parse > 'primitiveDate' 1`] = `"z.date()"`;

exports[`zod parse > 'primitiveDate' v4 1`] = `"z.date()"`;

exports[`zod parse > 'ref' 1`] = `"z.lazy(() => Pet)"`;

exports[`zod parse > 'ref' v4 1`] = `"Pet"`;

exports[`zod parse > 'string' 1`] = `"z.string()"`;

exports[`zod parse > 'string' v4 1`] = `"z.string()"`;

exports[`zod parse > 'stringLocal' 1`] = `"z.string().datetime({ local: true })"`;

exports[`zod parse > 'stringLocal' v4 1`] = `"z.iso.datetime({ local: true })"`;

exports[`zod parse > 'stringOffset' 1`] = `"z.string().datetime({ offset: true })"`;

exports[`zod parse > 'stringOffset' v4 1`] = `"z.iso.datetime({ offset: true })"`;

exports[`zod parse > 'time' 1`] = `"z.string().time()"`;

exports[`zod parse > 'time' v4 1`] = `"z.iso.time()"`;

exports[`zod parse > 'tuple' 1`] = `"z.tuple([])"`;

exports[`zod parse > 'tuple' v4 1`] = `"z.tuple([])"`;

exports[`zod parse > 'tupleMulti' 1`] = `"z.tuple([z.string(), z.number()])"`;

exports[`zod parse > 'tupleMulti' v4 1`] = `"z.tuple([z.string(), z.number()])"`;

exports[`zod parse > 'undefined' 1`] = `"z.undefined()"`;

exports[`zod parse > 'undefined' v4 1`] = `"z.undefined()"`;

exports[`zod parse > 'union' 1`] = `"z.union([z.string(), z.number()])"`;

exports[`zod parse > 'union' v4 1`] = `"z.union([z.string(), z.number()])"`;

exports[`zod parse > 'unionOne' 1`] = `"z.string()"`;

exports[`zod parse > 'unionOne' v4 1`] = `"z.string()"`;

exports[`zod parse > 'unknown' 1`] = `"z.unknown()"`;

exports[`zod parse > 'unknown' v4 1`] = `"z.unknown()"`;
